### PR TITLE
[Feature] Support Class Aware Sampler

### PR DESCRIPTION
**Pull request by BIGWangYuDong**
continue  #7210

This PR is based on https://github.com/open-mmlab/mmdetection/pull/6331.

More details will be updated when this PR is fully prepared

### CODE REVIEW

File: configs/openimages/README.md [hunk 1]
```diff
@@ -90,14 +93,14 @@ training/testing by using `tools/misc/get_image_metas.py`.
     │   │   │   ├── class-descriptions-boxable.csv
     │   │   │   ├── oidv6-train-annotations-bbox.scv
     │   │   │   ├── validation-annotations-bbox.csv
-    │   │   │   ├── validation-annotations-human-imagelabels-boxable.csv    # is not necessary
+    │   │   │   ├── validation-annotations-human-imagelabels-boxable.csv    # should set `load_image_level_labels=False` if not use
```
Comment (ZwwWayne) [inline configs/openimages/README.md:96]: 1. set load_image_level_label=False in which if not use?
2. Should leave this not in other place.

File: mmdet/datasets/builder.py [hunk 2]
```diff
@@ -93,6 +92,7 @@ def build_dataloader(dataset,
                      seed=None,
                      runner_type='EpochBasedRunner',
                      persistent_workers=False,
+                     use_class_aware_sampler=None,
```
Comment (ZwwWayne) [inline mmdet/datasets/builder.py:95]: use_class_aware_sampler=None, -> class_aware_sampler=None,

File: mmdet/datasets/samplers/class_aware_sampler.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import math
+
+import torch
+from mmcv.runner import get_dist_info
+from torch.utils.data import Sampler
+
+
+class ClassAwareSampler(Sampler):
+    r"""A class-aware sampling strategy to effectively tackle the
```
Comment (ZwwWayne) [inline mmdet/datasets/samplers/class_aware_sampler.py:10]: The summary should be short and precise.

Also describe whether it can be used in distributed or non-distributed scenarios.

File: mmdet/datasets/samplers/class_aware_sampler.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        self.epoch = 0
+        self.seed = seed if seed is not None else 0
+
+        # The number of samples taken from each per-label list
+        assert num_sample_class > 0 and isinstance(num_sample_class, int)
+        self.num_sample_class = num_sample_class
+        # Get per-label image list from dataset
+        assert hasattr(dataset, 'get_label_dict'), \
+            'dataset must have `get_label_dict` function'
+        self.class_dict = dataset.get_label_dict()
```
Comment (ZwwWayne) [inline mmdet/datasets/samplers/class_aware_sampler.py:65]: The data set should already have a common interface named get_cat_ids.

File: mmdet/datasets/samplers/class_aware_sampler.py (hunk 3 above)
Comment (BIGWangYuDong) [reply mmdet/datasets/samplers/class_aware_sampler.py:10]: Seems this sampler can be used in both dist and non-dist scenarios

File: mmdet/datasets/samplers/class_aware_sampler.py (hunk 4 above)
Comment (BIGWangYuDong) [reply mmdet/datasets/samplers/class_aware_sampler.py:65]: They are not the same, `get_cat_ids` is based on images, which will return a [image_idx[label idx]]. However, `get_label_dict` is based on label, which will return a dict[label_idx[image_idx]]

File: configs/openimages/README.md [hunk 5]
```diff
@@ -112,14 +115,33 @@ Open Images v6, but the test images are different.
 You can also download the annotations from [official website](https://storage.googleapis.com/openimages/web/challenge2019_downloads.html),
 and set data.train.type=OpenImagesDataset, data.val.type=OpenImagesDataset, and data.test.type=OpenImagesDataset in the config
 3. If users do not want to use `validation-annotations-human-imagelabels-boxable.csv` and `challenge-2019-validation-detection-human-imagelabels.csv`
-users can should set `data.val.load_image_level_labels=False` and `data.test.load_image_level_labels=False` in the config .
+users can set `data.val.load_image_level_labels=False` and `data.test.load_image_level_labels=False` in the config. Loading image-levels label is the
+Open Images [evaluation metric](https://storage.googleapis.com/openimages/web/evaluation.html). If the images are annotated with positive image-level labels, indicating certain object classes are present, and with negative image-level labels, indicating certain classes are absent.
+All other classes are unannotated.
+
+
```
Comment (ZwwWayne) [inline configs/openimages/README.md:122]: unnecessary empty line.

File: configs/openimages/README.md [hunk 6]
```diff
@@ -112,14 +115,33 @@ Open Images v6, but the test images are different.
 You can also download the annotations from [official website](https://storage.googleapis.com/openimages/web/challenge2019_downloads.html),
 and set data.train.type=OpenImagesDataset, data.val.type=OpenImagesDataset, and data.test.type=OpenImagesDataset in the config
 3. If users do not want to use `validation-annotations-human-imagelabels-boxable.csv` and `challenge-2019-validation-detection-human-imagelabels.csv`
-users can should set `data.val.load_image_level_labels=False` and `data.test.load_image_level_labels=False` in the config .
+users can set `data.val.load_image_level_labels=False` and `data.test.load_image_level_labels=False` in the config. Loading image-levels label is the
+Open Images [evaluation metric](https://storage.googleapis.com/openimages/web/evaluation.html). If the images are annotated with positive image-level labels, indicating certain object classes are present, and with negative image-level labels, indicating certain classes are absent.
```
Comment (ZwwWayne) [inline configs/openimages/README.md:119]: rephrase the sentense.

File: mmdet/datasets/custom.py [hunk 7]
```diff
@@ -285,6 +285,28 @@ def get_classes(cls, classes=None):
...
+        used in :class:`ClassAwareSampler`.
+
+        Returns:
+            dict[list]: A ddict of per-label image list,
+            the item of the dict indicates a label index,
+            corresponds to the image index that contains the label.
+        """
+        label_dict = dict()
+        if self.CLASSES is None:
+            raise ValueError('CLASSES can not be None')
```
Comment (ZwwWayne) [inline mmdet/datasets/custom.py:299]: self.CLASSES can not be

File: mmdet/datasets/custom.py [hunk 8]
```diff
@@ -285,6 +285,28 @@ def get_classes(cls, classes=None):
...
+        # sort the label index
+        for i in range(len(self.CLASSES)):
+            label_dict[i] = []
```
Comment (ZwwWayne) [inline mmdet/datasets/custom.py:302]: merge line 301/302 & 297 into
```
label_dict -> label2imgs
```

File: mmdet/datasets/custom.py [hunk 9]
```diff
@@ -285,6 +285,28 @@ def get_classes(cls, classes=None):
 
         return class_names
 
+    def get_index_dict(self):
```
Comment (ZwwWayne) [inline mmdet/datasets/custom.py:288]: get_index_dict -> get_imgs_of_label

Comment (ZwwWayne) [reply mmdet/datasets/custom.py:288]: get_index_dict -> get_cat2imgs

File: mmdet/datasets/custom.py [hunk 10]
```diff
@@ -285,6 +285,28 @@ def get_classes(cls, classes=None):
...
+            the item of the dict indicates a label index,
+            corresponds to the image index that contains the label.
+        """
+        label_dict = dict()
+        if self.CLASSES is None:
+            raise ValueError('CLASSES can not be None')
+        # sort the label index
+        for i in range(len(self.CLASSES)):
+            label_dict[i] = []
+        data_infos = [self.get_ann_info(idx) for idx in range(len(self))]
```
Comment (ZwwWayne) [inline mmdet/datasets/custom.py:303]: we can use self.get_cat_ids here, consider:
```
for i in range(len(self)):
    cat_ids = set(self.get_cat_ids)
    for cat in cat_ids:
        cat2imgs[cat].append(i)
```

File: mmdet/datasets/samplers/class_aware_sampler.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+            num_replicas = _num_replicas
+        if rank is None:
+            rank = _rank
+
+        self.dataset = dataset
+        self.num_replicas = num_replicas
+        self.samples_per_gpu = samples_per_gpu
+        self.rank = rank
+        self.epoch = 0
+        self.seed = seed if seed is not None else 0
```
Comment (ZwwWayne) [inline mmdet/datasets/samplers/class_aware_sampler.py:57]: Should use sync_random_seed, see those in DistributedSampler.

File: mmdet/datasets/samplers/class_aware_sampler.py [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+        assert hasattr(dataset, 'get_index_dict'), \
+            'dataset must have `get_index_dict` function'
+        self.class_dict = dataset.get_index_dict()
+
+        self.num_samples = int(
+            math.ceil(
+                len(self.dataset) * 1.0 / self.num_replicas /
+                self.samples_per_gpu)) * self.samples_per_gpu
+        self.total_size = self.num_samples * self.num_replicas
+        self.class_num = len(self.class_dict.keys())
```
Comment (ZwwWayne) [inline mmdet/datasets/samplers/class_aware_sampler.py:72]: here we can all use cat rather than class.

File: configs/openimages/README.md [hunk 13]
```diff
@@ -1,6 +1,9 @@
 # Open Images Dataset
-<!-- [DATASET] -->
 
+> [Open Images Dataset](https://arxiv.org/abs/1811.00982)
+
+<!-- [DATASET] -->
+<!-- [ALGORITHM] -->
```
Comment (jbwang1997) [inline configs/openimages/README.md:6]: We should assign one label to this config.

File: mmdet/datasets/custom.py [hunk 14]
```diff
@@ -285,6 +285,25 @@ def get_classes(cls, classes=None):
...
+            dict[list]: A dict of per-label image list,
+            the item of the dict indicates a label index,
+            corresponds to the image index that contains the label.
+        """
+        if self.CLASSES is None:
+            raise ValueError('self.CLASSES can not be None')
+        # sort the label index
+        cat2imgs = {i: [] for i in range(len(self.CLASSES))}
+        for i in range(len(self)):
+            cat_ids = set(self.get_cat_ids(i))
```
Comment (jbwang1997) [inline mmdet/datasets/custom.py:302]: Could we replace `i` with a specific function name?

File: mmdet/datasets/openimages.py [hunk 15]
```diff
@@ -563,6 +563,17 @@ def denormalize_gt_bboxes(self, annotations):
...
+    def get_cat_ids(self, idx):
+        """Get category ids by index.
+
+        Args:
+            idx (int): Index of data.
+
+        Returns:
+            list[int]: All categories in the image of specified index.
+        """
+        return self.get_ann_info(idx)['labels'].astype(np.int).tolist()
```
Comment (jbwang1997) [inline mmdet/datasets/openimages.py:575]: Seems `OpenImagesDataset` inherit `CustomDataset`. Is this part necessary?

File: mmdet/datasets/samplers/class_aware_sampler.py [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+        if self.i < self.length:
+            idx = self.data[self.index[self.i]]
+            self.i += 1
+            return idx
+        else:
+            self.index = torch.randperm(
+                self.length, generator=self.generator).numpy()
+            self.i = 1
+            idx = self.data[self.index[self.i - 1]]
+            return idx
```
Comment (jbwang1997) [inline mmdet/datasets/samplers/class_aware_sampler.py:179]:
```
self.i %= len(self)
if self.i == 0:
    self.index = torch.randperm(self.length, generator=self.generator).numpy()

idx = self.data[self.index[self.i]]
self.i += 1
```
Could we simplify the logic like this?

Comment (jbwang1997) [reply mmdet/datasets/samplers/class_aware_sampler.py:179]: I consider it's more elegant to apply `torch.randperm` in one place.

**Comment by BIGWangYuDong** [comment]
we need update codes after https://github.com/open-mmlab/mmdetection/pull/7668 has merged.

File: mmdet/datasets/openimages.py (hunk 15 above)
Comment (BIGWangYuDong) [reply mmdet/datasets/openimages.py:575]: Yes, because of the keys are different, there are some differences between them: 
In OID, we use 
`self.get_ann_info(idx)['labels'].astype(np.int).tolist()` 
but in CustomDataset, we use 
`self.data_infos[idx]['ann']['labels'].astype(np.int).tolist()`

File: mmdet/datasets/samplers/class_aware_sampler.py [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+        self.generator = generator
+
+    def __iter__(self):
+        return self
+
+    def __len__(self):
+        return len(self.data)
+
+    def __next__(self):
+        self.i %= (self.length + 1)
```
Comment (jbwang1997) [inline mmdet/datasets/samplers/class_aware_sampler.py:170]: `self.i %= (self.length + 1)` is unnecessary.

File: configs/openimages/README.md (hunk 13 above)
Comment (ZwwWayne) [reply configs/openimages/README.md:6]: Just use Dataset is fine

File: mmdet/datasets/pipelines/loading.py [hunk 18]
```diff
@@ -256,16 +256,21 @@ def _load_bboxes(self, results):
...
             if bbox_num != 0:
+                h, w = results['img_shape'][:2]
                 results['gt_bboxes'][:, 0::2] *= w
                 results['gt_bboxes'][:, 1::2] *= h
-            results['gt_bboxes'] = results['gt_bboxes'].astype(np.float32)
 
         gt_bboxes_ignore = ann_info.get('bboxes_ignore', None)
         if gt_bboxes_ignore is not None:
             results['gt_bboxes_ignore'] = gt_bboxes_ignore.copy()
+            if self.denorm_bbox:
```
Comment (ZwwWayne) [inline mmdet/datasets/pipelines/loading.py:268]: Seems that unnecessary modification is involved, please check whether we need them

File: mmdet/datasets/pipelines/loading.py [hunk 19]
```diff
@@ -256,12 +256,11 @@ def _load_bboxes(self, results):
         results['gt_bboxes'] = ann_info['bboxes'].copy()
 
         if self.denorm_bbox:
-            h, w = results['img_shape'][:2]
             bbox_num = results['gt_bboxes'].shape[0]
             if bbox_num != 0:
+                h, w = results['img_shape'][:2]
                 results['gt_bboxes'][:, 0::2] *= w
                 results['gt_bboxes'][:, 1::2] *= h
-            results['gt_bboxes'] = results['gt_bboxes'].astype(np.float32)
```
Comment (BIGWangYuDong) [inline mmdet/datasets/pipelines/loading.py:264]: the default type is np.float32, so delete this line

**Review by jbwang1997 — APPROVED** [review]
LGTM

File: mmdet/apis/train.py [hunk 20]
```diff
@@ -129,7 +129,6 @@ def train_detector(model,
 
     runner_type = 'EpochBasedRunner' if 'runner' not in cfg else cfg.runner[
         'type']
-
```
Comment (ZwwWayne) [inline mmdet/apis/train.py:132]: seems to be unnecessary modification

File: mmdet/datasets/builder.py [hunk 21]
```diff
@@ -198,4 +211,3 @@ def worker_init_fn(worker_id, num_workers, rank, seed):
     worker_seed = num_workers * rank + worker_id + seed
     np.random.seed(worker_seed)
     random.seed(worker_seed)
-    torch.manual_seed(worker_seed)
```
Comment (ZwwWayne) [inline mmdet/datasets/builder.py:201]: Seems to be irrelevant modification?

File: mmdet/datasets/samplers/class_aware_sampler.py [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+        assert hasattr(dataset, 'get_cat2imgs'), \
+            'dataset must have `get_cat2imgs` function'
+        self.cat_dict = dataset.get_cat2imgs()
+
+        self.num_samples = int(
+            math.ceil(
+                len(self.dataset) * 1.0 / self.num_replicas /
+                self.samples_per_gpu)) * self.samples_per_gpu
+        self.total_size = self.num_samples * self.num_replicas
+        self.cat_num = len(self.cat_dict.keys())
```
Comment (ZwwWayne) [inline mmdet/datasets/samplers/class_aware_sampler.py:77]: self.cat_num -> self.num_cat_imgs
also add comments to indicate it means number of images containing each category.

File: mmdet/datasets/samplers/class_aware_sampler.py [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+        self.cat_num_list = [
+            len(self.cat_dict[i]) for i in range(self.cat_num)
+        ]
+        # filter labels without images
+        self.cat_index = [
```
Comment (ZwwWayne) [inline mmdet/datasets/samplers/class_aware_sampler.py:82]: self.cat_inds

File: mmdet/datasets/samplers/class_aware_sampler.py [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+            `num_gpus * samples_per_gpu`.
+            Default : 1.
+        num_replicas (optional): Number of processes participating in
+            distributed training.
+        rank (optional): Rank of the current process within num_replicas.
+        seed (int, optional): random seed used to shuffle the sampler if
+            ``shuffle=True``. This number should be identical across all
+            processes in the distributed group. Default: 0.
+        num_sample_class (int): The number of samples taken from each
+            per-label list, which is used in :class:`ClassAwareSampler`.
```
Comment (ZwwWayne) [inline mmdet/datasets/samplers/class_aware_sampler.py:37]: 'which is used in :class:`ClassAwareSampler`.' is an unnecessary description

File: mmdet/datasets/samplers/class_aware_sampler.py [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+        g.manual_seed(self.epoch + self.seed)
+
+        # initialize label list
+        label_iter_list = RandomCycleIter(self.cat_index, generator=g)
+        # initialize each per-label image list
+        data_iter_dict = dict()
+        for i in self.cat_index:
+            data_iter_dict[i] = RandomCycleIter(self.cat_dict[i], generator=g)
+
+        def gen_cat_num_indices(cls_list, data_dict, num_sample_cls):
```
Comment (ZwwWayne) [inline mmdet/datasets/samplers/class_aware_sampler.py:99]: gen_cat_num_indices is not a very clear name, does it mean generate indices for each category?

File: mmdet/datasets/samplers/class_aware_sampler.py (hunk 23 above)
Comment (ZwwWayne) [reply mmdet/datasets/samplers/class_aware_sampler.py:82]: or self.valid_cat_inds

File: mmdet/datasets/samplers/class_aware_sampler.py [hunk 26]
```diff
@@ -0,0 +1,179 @@
...
+
+        self.num_samples = int(
+            math.ceil(
+                len(self.dataset) * 1.0 / self.num_replicas /
+                self.samples_per_gpu)) * self.samples_per_gpu
+        self.total_size = self.num_samples * self.num_replicas
+        self.cat_num = len(self.cat_dict.keys())
+
+        # get number of images containing each category
+        self.num_cat_imgs = [
```
Comment (ZwwWayne) [inline mmdet/datasets/samplers/class_aware_sampler.py:79]:
```
        self.num_cat_imgs = [
            len(x) for x in self.cat_dict.items()
        ]
        # filter labels without images
        self.valid_cat_inds = [
            i for i, length in enumerate(self.num_cat_imgs) if length != 0
        ]
        self.num_classes = len(self.valid_cat_inds)
```

File: mmdet/datasets/samplers/class_aware_sampler.py [hunk 27]
```diff
@@ -0,0 +1,179 @@
...
+        assert hasattr(dataset, 'get_cat2imgs'), \
+            'dataset must have `get_cat2imgs` function'
+        self.cat_dict = dataset.get_cat2imgs()
+
+        self.num_samples = int(
+            math.ceil(
+                len(self.dataset) * 1.0 / self.num_replicas /
+                self.samples_per_gpu)) * self.samples_per_gpu
+        self.total_size = self.num_samples * self.num_replicas
+        self.cat_num = len(self.cat_dict.keys())
```
Comment (ZwwWayne) [inline mmdet/datasets/samplers/class_aware_sampler.py:76]: line 76 can be deleted.
